Add a nbsp at the the end of the text when the next text has a leading space

Ahmad-S792 opened the pull request:
#### 41ea6427c93797b6c7aceb36d05329618ddd87b9
<pre>
Add a nbsp at the the end of the text when the next text has a leading space

Add a nbsp at the the end of the text when the next text has a leading space
<a href="https://bugs.webkit.org/show_bug.cgi?id=248723">https://bugs.webkit.org/show_bug.cgi?id=248723</a>

Reviewed by Ryosuke Niwa.

This patch is to fix three regressions stemming from 257136@main.

1) Breaking chat window or login forms while trying to press &apos;space&apos; key.
2) Add nbsp instead of plain space between text nodes in case next text node has a leading plain space.
3) Not accounting for all kind of whitespaces including &quot;Enter&quot;

These are merge of following Blink commits:

&gt; <a href="https://chromium.googlesource.com/chromium/src.git/+/4dab74137593abb0888e415294aeb80da27362e3">https://chromium.googlesource.com/chromium/src.git/+/4dab74137593abb0888e415294aeb80da27362e3</a>
&gt; <a href="https://chromium.googlesource.com/chromium/src.git/+/9ebff223d5d639167c3efb01a5fa912b3aee05a9">https://chromium.googlesource.com/chromium/src.git/+/9ebff223d5d639167c3efb01a5fa912b3aee05a9</a>
&gt; <a href="https://chromium.googlesource.com/chromium/src.git/+/5b9a9a6f8c9c646ac542a9e1a2f18c59140849e6">https://chromium.googlesource.com/chromium/src.git/+/5b9a9a6f8c9c646ac542a9e1a2f18c59140849e6</a>

* Source/WebCore/editing/CompositeEditingCommand.cpp:
(CompositeEditCommand::rebalanceWhitespaceOnTextSubtring):
(1) Add logic to account for leading space (including &quot;Enter&quot; key)
(2) accept &quot;Space&quot; Key
(3) Refactor code by introducing const bool &quot;nextSiblingIsTextNodeWithoutLeadingSpace&quot;
* LayoutTests/resources/assert-selection.js: Added Test Script with Chromium Copyright header
* LayoutTests/editing/inserting/insert-space.html: Added Test Case
* LayoutTests/editing/inserting/insert-space-expected.txt: Added Test Case Expectation

Canonical link: <a href="https://commits.webkit.org/257622@main">https://commits.webkit.org/257622@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d6fb9659ce1b03c2296efb5bd67cde6545be87c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99448 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8633 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32550 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108819 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169054 "Failed to checkout and rebase branch from PR 7257") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103440 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9204 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85942 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91924 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106741 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105205 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6960 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90490 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33925 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88779 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21837 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2487 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23354 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2407 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5242 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8564 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42829 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4274 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/CompositeEditCommand.cpp:
```diff
@@ -959,9 +959,9 @@ void CompositeEditCommand::rebalanceWhitespaceOnTextSubstring(Text& textNode, in
     String string = text.substring(upstream, length);
     // FIXME: Because of the problem mentioned at the top of this function, we must also use nbsps at the start/end of the string because
     // this function doesn't get all surrounding whitespace, just the whitespace in the current text node.
+    const bool nextSiblingIsTextNode = textNode.nextSibling() && textNode.nextSibling()->isTextNode() && downcast<Text>(textNode.nextSibling())->data().length() && downcast<Text>(textNode.nextSibling())->data()[0] != ' ';
```
I'd rather call this variable nextSiblingIsTextNodeWithoutLeadingSpace to be more explicit.

Ahmad-S792 replied:
Fixed but EWS hook is not picking the latest change yet.

rniwa reviewed (approved):
r=me

Ahmad-S792 committed:
Add a nbsp at the the end of the text when the next text has a leading space

Add a nbsp at the the end of the text when the next text has a leading space
https://bugs.webkit.org/show_bug.cgi?id=248723

Reviewed by Ryosuke Niwa.

This patch is to fix three regressions stemming from 257136@main.

1) Breaking chat window or login forms while trying to press 'space' key.
2) Add nbsp instead of plain space between text nodes in case next text node has a leading plain space.
3) Not accounting for all kind of whitespaces including "Enter"

These are merge of following Blink commits:

> https://chromium.googlesource.com/chromium/src.git/+/4dab74137593abb0888e415294aeb80da27362e3
> https://chromium.googlesource.com/chromium/src.git/+/9ebff223d5d639167c3efb01a5fa912b3aee05a9
> https://chromium.googlesource.com/chromium/src.git/+/5b9a9a6f8c9c646ac542a9e1a2f18c59140849e6

* Source/WebCore/editing/CompositeEditingCommand.cpp:
(CompositeEditCommand::rebalanceWhitespaceOnTextSubtring):
(1) Add logic to account for leading space (including "Enter" key)
(2) accept "Space" Key
(3) Refactor code by introducing const bool "nextSiblingIsTextNodeWithoutLeadingSpace"
* LayoutTests/resources/assert-selection.js: Added Test Script with Chromium Copyright header
* LayoutTests/editing/inserting/insert-space.html: Added Test Case
* LayoutTests/editing/inserting/insert-space-expected.txt: Added Test Case Expectation

Canonical link: https://commits.webkit.org/257622@main
